packages/markdown-support/src/codeblock.ts

```diff
@@ -1,43 +1,44 @@
 import { visit } from 'unist-util-visit';
+import type { Node } from 'unist';
+import type { Element } from 'hast';
+import type { Code } from 'mdast';
 
 /**  */
 export function remarkCodeBlock() {
-  const visitor = (node: any) => {
-    const { data, meta } = node;
-    let lang = node.lang || 'html'; // default to html matches GFM behavior.
+  return function (tree: Node) {
+    visit(tree, 'code', (node: Code) => {
+      const { data, meta } = node;
+      let lang = node.lang || 'html'; // default to html matches GFM behavior.
 
-    let currentClassName = data?.hProperties?.class ?? '';
-    node.data = node.data || {};
-    node.data.hProperties = node.data.hProperties || {};
-    node.data.hProperties = { ...node.data.hProperties, class: `language-${lang} ${currentClassName}`.trim(), lang, meta };
-
-    return node;
+      let currentClassName = (data?.hProperties as { [k: string]: string })?.class ?? '';
```

Comment:
I couldn't avoid typecasting here 🙈 `data` is typed as `[key: string]: unknown` in unist, and I couldn't think of a better solution.

Please suggest any better ideas!

Reply:
So, lots of background first:
Similar to what you had with `remark-autolink-headings` and `rehype-autolink-headings`, while a lot is possible by defining what markdown things turn into when they are turned to HTML (through `hProperties`), a lot is easier if you’re working on the HTML instead!

Markdown is *not* just about HTML. For example, it has two different link types. If you care about such things, do stuff in remark. If you‘re only really caring about HTML stuff, do things in remark.

As an analogy, say you could do things in a TS AST, and in a JS AST. If you do things in the TS AST, you could do some extra things, but then JS folks can’t use your plugin. If you do something in a JS AST you don’t need to worry about TS specific annotations, and more users could use your plugin!

So, most of what’s done here is makes more sense in rehype/hast/HTML space.
But, it might be a bit much to change that now.
So backtracking: here we’re defining in markdown, what stuff turns into when it’s turned to HTML.
So, in mdast, we patch some hast stuff.

`hProperties` is a semistandard field handled by a few plugins.
It’s used to add the things defined in it as properties to an element.
The proper interface for it is `Properties` from `@types/hast`!
Oh and also: this works, mostly, but it should be called `className` and set to an array of strings!

Reply:
Thank you for the background 🙌 In our case, I'm actually not sure why we have this custom CodeBlock plugin in both remark and rehype, maybe we can only have it in rehype. Do you know @matthewp?

But I agree we might want to discuss/tackle this separately to avoid this getting too out of hands 🙂 

---

> Oh and also: this works, mostly, but it should be called className and set to an array of strings!

Do you mean something like this?:
```diff
- let currentClassName = (data?.hProperties as Properties)?.class ?? '';
+ let currentClassName = (data?.hProperties as Properties)?.className?? [''];
node.data = node.data || {};
node.data.hProperties = node.data.hProperties || {};
- node.data.hProperties = { ...(node.data.hProperties as Properties), class: `language-${lang} ${currentClassName}`.trim(), lang, meta };
+ node.data.hProperties = { ...(node.data.hProperties as Properties), className: `language-${lang} ${currentClassName.join(' ')}`.trim(), lang, meta };
```

(seems like it, I've found docs on this over [here](https://github.com/syntax-tree/hast#propertyvalue))

For some reason, this breaks a test about MD syntax highlighting, but maybe the test is set up differently, I'm not really familiar with the Astro specs yet

Reply:
About the `class` vs `className` shenanigans ☝️ 

I was able to see the problem by nesting a markdown code block in a `<div>` (same setup as the tests):

```html
<div>
    ```js
    import Something from './another';

    const thing = new Something();
    ```
</div>
```

Looks like for us, we still want to keep `class`, because this is what the `node` looks like in the plugin:

```js
{
  type: 'code',
  lang: 'js',
  meta: null,
  value: "import Something from './another';\n\nconst thing = new Something();",
  position: {
    start: { line: 2, column: 1, offset: 1 },
    end: { line: 6, column: 4, offset: 77 }
  },
  data: { hProperties: { class: 'astro-ldKR6c5M' } } // 👈 it's `class`
}
```

It seems like this is actually set by one of our other unified plugins, [remark-scoped-styles](https://github.com/snowpackjs/astro/blob/3bfd8c125e57879f6d0d3a828943a16199c8fb9c/packages/markdown-support/src/remark-scoped-styles.ts).

I suggest keeping `class` for now, and we can look into amending it to match [HAST properties](https://github.com/syntax-tree/hast#properties) in a future PR